pkg/drivers/docker_driver.go

```diff
@@ -0,0 +1,371 @@
+}
+
+func (d *DockerDriver) ReadFile(target string) ([]byte, error) {
+	reader, err := d.retrieveTar(target)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if filepath.Clean(header.Name) == path.Base(target) {
+				return nil, fmt.Errorf("Cannot read specified path: %s is a directory, not a file", target)
+			}
+		case tar.TypeSymlink:
+			return d.ReadFile(header.Linkname)
+		case tar.TypeReg, tar.TypeLink:
+			if filepath.Clean(header.Name) == path.Base(target) {
+				var b bytes.Buffer
+				stream := bufio.NewWriter(&b)
+				io.Copy(stream, reader)
+				return b.Bytes(), nil
+			}
+		default:
+			continue
+		}
+	}
+	return nil, fmt.Errorf("File %s not found in image", target)
+}
+
+func (d *DockerDriver) ReadDir(target string) ([]os.FileInfo, error) {
+	reader, err := d.retrieveTar(target)
+	if err != nil {
+		return nil, err
+	}
+	var infos []os.FileInfo
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if header.Typeflag == tar.TypeDir {
+			// we only want top level dirs here, no recursion. to get these, remove
+			// trailing separator and split on separator. there should only be two parts.
+			parts := strings.Split(strings.TrimSuffix(header.Name, string(os.PathSeparator)), string(os.PathSeparator))
+			if len(parts) == 2 {
+				infos = append(infos, header.FileInfo())
+			}
+		}
+	}
+	return infos, nil
+}
+
+// This method takes a command (in the form of a list of args), and does the following:
+// 1) creates a container, based on the "current latest" image, with the command set as
+// the command to run when the container starts
+// 2) starts the container
+// 3) commits the container with its changes to a new image,
+// and sets that image as the new "current image"
+func (d *DockerDriver) runAndCommit(env []string, command []string) string {
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Env:          env,
+			Cmd:          command,
+			Entrypoint:   []string{},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
+		return ""
+	}
+
+	err = d.cli.StartContainer(container.ID, nil)
+	if err != nil {
+		logrus.Errorf("Error creating container: %s", err.Error())
+	}
+
+	_, err = d.cli.WaitContainer(container.ID)
+
+	image, err := d.cli.CommitContainer(docker.CommitContainerOptions{
+		Container: container.ID,
+	})
+
+	if err != nil {
+		logrus.Errorf("Error committing container: %s", err.Error())
+	}
+
+	if !d.save {
+		if err = d.cli.RemoveContainer(docker.RemoveContainerOptions{
+			ID: container.ID,
+		}); err != nil {
+			logrus.Infof("Error when removing container %s: %s", container.ID, err.Error())
+		}
+	}
+
+	d.currentImage = image.ID
```

Comment:
Will this work if we call it concurrently? If not, maybe a note that this function should never be called concurrently.

Reply:
depends on how the concurrency was set up. if we just threw a bunch of threads at this and said "go" this could definitely fail, especially if the steps to the tests were processed out of order. but if we sent each test on its own goroutine and had them run in parallel, they should actually work as intended assuming the steps of each test were run in order. in that sense, I think it would actually be pretty hard to call this function in a dangerous way.